source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,467 @@
+
+Accessors defined by the Interface
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The interface we implement specifies some methods to access ``Value`` objects. For ``HealthData``, those are
+``health()`` and ``maxHealth()``. Every call to those should yield a new ``Value``.
+
+.. code-block:: java
+
+    public MutableBoundedValue<Double> health() {
+        return new SpongeBoundedValue<>(Keys.HEALTH, // 1
+            this.maximumHealth,  // 2
+            ComparatorUtil.doubleComparator(), // 3
+            0D, // 4
+            (double) Float.MAX_VALUE, // 5
+            this.currentHealth); // 6
+    }
+
+Since we use a bounded value, our constructor is slightly longer.
+
+.. csv-table::
+    :header: "Number", "Parameter", "Needed for"
+    :widths: 4, 40, 40
+
+    "1", "``Key`` identifying the value", "every subclass of ``BaseValue``"
+    "2", "the default value", "every subclass of ``BaseValue``"
+    "3", "the ``Comparator`` to use", "bounded values"
+    "4", "the minimum acceptable value", "bounded values"
+    "5", "the maximum acceptable value", "bounded values"
+    "6", "the current value", "every subclass of ``BaseValue``, can be omitted"
+
+If no current value is specified, calling ``get()`` on the ``Value`` returns the default value.
+
+Copying and Serialization
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The two methods ``copy()`` and ``asImmutable()`` are not much work to implement. For both you just need to return
+a mutable or an immutable data manipulator respectively, containing the same data as the current instance.
+
+The method ``toContainer()`` is used for serialization purposes. Use a ``MemoryDataContainer`` as the result
+and apply to it the values stored within this instance. A ``DataContainer`` is basically a map mapping ``DataQuery``\ s
+to values. Since a ``Key`` always contains a corresponding ``DataQuery``, just use those by passing the ``Key`` directly.
+
+.. code-block:: java
+
+    public DataContainer toContainer() {
+        return new MemoryDataContainer()
+            .set(Keys.HEALTH, this.currentHealth)
+            .set(Keys.MAX_HEALTH, this.maximumHealth);
+    }
+
+registerGettersAndSetters()
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A ``DataManipulator`` also provides methods to get and set data using keys. The implementation for this is handled
+by ``AbstractData``, but we must tell it which data it can access and how. Therefore, in the
+``registerGettersAndSetters()`` method we need to do the following for each value:
+
+* register a ``Supplier`` to directly get the value
+* register a ``Consumer`` to directly set the value
+* register a ``Supplier<Value>`` to get the mutable ``Value``
+
+``Supplier`` and ``Consumer`` are functional interfaces, so Java 8 Lambdas can be used.
+
+.. code-block:: java
+
+ private void registerGettersAndSetters() {
+      registerFieldGetter(Keys.HEALTH, () -> SpongeHealthData.this.currentHealth);
+      registerFieldSetter(Keys.HEALTH, value -> SpongeHealthData.this.currentHealth = value);
```

Comment:
I would argue that here needs to be a little clarified, specifically that the setter should likely validate values (non-nulls) depending on the data being inputted. I took a brief looks at most of the manipulators in action and noticed on my end that most of the manipulators do NOT perform this type of validation (especially ones where you shouldn't be setting negative number values).

That said, using lambdas are fine though personally, I'd still use a method to do such a thing, makes things a little more easier to read.